File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
def makeData(str):
    lst = []
    for i in range(0, len(str)-1):
        if str[i].isalpha() and str[i+1].isalpha():
            lst.append(str[i].lower()+str[i+1].lower())
    return lst

def countElement(lst):
    newDict = {}
    for i in lst:
        if i in newDict:
            newDict[i] += 1
        else:
            newDict[i] = 1
    return newDict

def jaccard(dicA, dicB, intersectionList, unionList):
    cntIntersection = 0
    cntUnion = 0
    print(dicA, dicB, intersectionList, unionList)
    for intersec in intersectionList:
        cntIntersection += min(dicA[intersec], dicB[intersec])
    for uni in unionList:
        if uni in dicA and uni in dicB:
            # 둘 다 있는 경우...
            cntUnion += max(dicA[uni], dicB[uni])
        elif uni in dicA:
            cntUnion += dicA[uni]
        else:
            cntUnion += dicB[uni]
    print(cntIntersection/cntUnion, '자카드', cntIntersection, cntUnion)
    return cntIntersection / cntUnion
    
    
    
def solution(str1, str2):
    answer = 0
    # J(A, B)는 두 집합의 교집합 크기를 두 집합의 합집합 크기로 나눈 값
    listA = makeData(str1)
    dictA = countElement(listA)
    listB = makeData(str2)
    dictB = countElement(listB)
    unionList = set(listA).union(listB)
    intersectionList = set(listA).intersection(listB)
    if len(listA) == 0 and len(listB) == 0: #둘다 공집합인 경우
        answer = 1
    else:
        answer = jaccard(dictA, dictB, intersectionList, unionList)
    
    return int(answer * 65536)

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.js
function isAlphabet(character) {
    return /[a-zA-Z]/.test(character);
}
function intersection(arr1, arr2) {
    return Array.from(new Set(arr1.filter(item => arr2.includes(item))))
}
function union(arr1, arr2) {
    return Array.from(new Set(arr1.concat(arr2)))
}
function makeData(str) {
    let result = []
    for (let i = 0; i < str.length -1; i++) {
        if (isAlphabet(str[i]) && isAlphabet(str[i+1])) {
            result.push(str[i].toLowerCase() + str[i+1].toLowerCase())
        }
    }
    return result
}
function dataToDict(datas) {
    let dic = {}
    for (let data of datas) {
        if (data in dic) {
            dic[data] += 1
        } else {
            dic[data] = 1
        }
    }
    return dic
}
function solution(str1, str2) {
    let answer = 0;
    let data1 = makeData(str1);
    let dic1 = dataToDict(data1);
    let data2 = makeData(str2);
    let dic2 = dataToDict(data2);
    
    let unicnt = 0;
    let intercnt = 0;
    
    interList = intersection(data1, data2);
    uniList = union(data1, data2);
    if (uniList.length === 0) {
        return 65536;
    }
    for (inter of interList) {
        intercnt += Math.min(dic1[inter], dic2[inter]);
    }
    for (uni of uniList) {
        if (data1.includes(uni) && data2.includes(uni)) {
            unicnt += Math.max(dic1[uni], dic2[uni]);
        } else if (data1.includes(uni)) {
            unicnt += dic1[uni];
        } else {
            unicnt += dic2[uni];
        }
    }
    
    answer = parseInt(intercnt / unicnt * 65536);
    
    return answer;
}
